### PR TITLE
[ETCM-468] add JSON-RPC method eth-getProof

### LINKED ISSUE #350: Implement (non-) membership proof for an MPT element with a given key
For client applications (like light clients) this would be very helpful. There's an implementation of this functionality in Geth.

### PR DESCRIPTION
**Pull request by lemastero**
# Description

Add eth_getProof JSON-RPC method. [EIP-1186](https://eips.ethereum.org/EIPS/eip-1186)

# Proposed Solution

Uses getProof implemented in #857.

# Testing

### CODE REVIEW

File: src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala
Comment (dmitry-worker) [inline src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala]:
```suggestion
  "EthProofService" should "handle getStorageAt request" in new TestSetup {
```

**Review by dmitry-worker — COMMENTED** [review]
PR's looking good! 
But if I would upgrade something - maybe I'd upgrade the testing
+ add at least one failed scenario - to be sure that error matches an expected one (3 error cases were defined)
+ maybe test the route itself, mocking the logic - in order to validate `json` it gives

**Comment by dzajkowski** [comment]
Would it make sense to have the happy path merged into dev and work on the rest in a separate PR?

**Comment by bsuieric** [comment]
test

File: src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala
Comment (AnastasiiaL) [inline src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala]: a comma would be nice in here
```suggestion
        val ver = verifyProof[Array[Byte], Array[Byte]](trie.getRootHash, keyToFind, p)
```

**Comment by lemastero** [comment]
@dmitry-worker
> maybe test the route itself, mocking the logic - in order to validate json it gives

I have added sth like that in: https://github.com/input-output-hk/mantis/pull/869/commits/dfd4ec3efee92040e825ae7c418942d680e238a8

> add at least one failed scenario - to be sure that error matches an expected one (3 error cases were defined)

Each we could use more tests. Let's add it as separate PR. WDYT?

File: src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala
Comment (leo-bogastry) [inline src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala]: I guess the tests are running now and this TODO can be removed?

File: src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala
Comment (AnastasiiaL) [inline src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala]: should it be uncommented or deleted?

File: src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala
Comment (AnastasiiaL) [reply src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala]: no, this is something me and Bogdan are gonna fix

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (leo-bogastry) [inline src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala]: Just a typo in `storage entory`

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala:182]: wouldn't `BlockResolver` be usable here?

File: src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala]: a little bit of code golf:
```scala
      .toRight(JsonRpcError.InvalidParams(s"Block $number not found"))
```

I think it's equivalent and 1 (ONE!) line less ;)

WDYT?

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthJsonMethodsImplicits.scala
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/jsonrpc/EthJsonMethodsImplicits.scala]: nitpick: do you mind adding an explicit type here? implicits behave a bit less frantic when those are used.

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala]: would it be possible to name this param in a less clashing way?

File: src/main/scala/io/iohk/ethereum/proof/MptProofVerifier.scala
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/proof/MptProofVerifier.scala]: is this a 'test' only mechanism? or should it be available for 'prod' code also?

File: src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala
Comment (dzajkowski) [inline src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala]: would it be possible to use diffx `matchTo`? it's already in the project and gives a better output.

**Review by dzajkowski — CHANGES_REQUESTED** [review]

File: src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala]: BlockResolver is removed.

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala:182]: Yes. We spend today some time to reuse BlockResolver to replace resolveBlock in EthProofService and EthService (details in branch: https://github.com/input-output-hk/mantis/tree/getproof-wip there is lots of failing tests - mocking needs to be fixed).

It will require lots of changes in EthServiceSpec & will lead to massive conflicts. So with @AnastasiiaL @bsuieric we decided to allow the duplication and get back to this when EthService will be splitted.

File: src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/BlockResolver.scala]: I like that!. Changed.

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthJsonMethodsImplicits.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthJsonMethodsImplicits.scala]: Sure. Done.

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala]: Renamed `blockNumber` to `block`. WDYT ?

PS JSON-RPC spec uses `BlockNumber`: https://github.com/etclabscore/ethereum-json-rpc-specification/blob/master/openrpc.json#L717 as a name of type. We have BlockParam (that is more accurate as this can be number or description like latest).

File: src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala
Comment (lemastero) [reply src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala]: I was not aware of this. Thanks :)
Done.

File: src/main/scala/io/iohk/ethereum/proof/MptProofVerifier.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/proof/MptProofVerifier.scala]: Moved to tests.

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (bsuieric) [inline src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala]: can we find a better name for this function so it gives us a better idea of what is trying to achieve? Thinking of something like ` def (determine|get|evaluate)AccountProof`

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (bsuieric) [inline src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala]: Seems that this is just another constructor for ProofAccount. We could add the constructor in ProofAccount companion object, and remove the unnecessary helper method from the service

File: src/main/scala/io/iohk/ethereum/proof/MptProofVerifier.scala
Comment (bsuieric) [reply src/main/scala/io/iohk/ethereum/proof/MptProofVerifier.scala]: I agree, this is used only in T and it should be moved, Already talked with Piotr about this, I'll do it tomorrow

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala]: any particular reason to name the param 'result' instead of 'proofAccount'?

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala]: so for every request sent - the new service is created? even for the requests with the same data?

File: src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala
Comment (AnastasiiaL) [inline src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala]: shuffle is not needed here, right? we are testing all the values from 'input' anyway

File: src/test/scala/io/iohk/ethereum/proof/MptProofVerifier.scala
Comment (AnastasiiaL) [inline src/test/scala/io/iohk/ethereum/proof/MptProofVerifier.scala]: why Unit was chosen as a return type for a verified proof?

File: src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala
Comment (lemastero) [reply src/test/scala/io/iohk/ethereum/jsonrpc/EthProofServiceSpec.scala]: Replaced comment with assertion.

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala]: Fixed.

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala]: Renamed

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala]: Taken care of here. It is slightly bigger change: https://github.com/input-output-hk/mantis/pull/869/commits/e2f01c4e4736e90bef9a4c203ee7c24fb56bf2a7

File: src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala
Comment (lemastero) [reply src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala]: True. Removed.

File: src/test/scala/io/iohk/ethereum/proof/MptProofVerifier.scala
Comment (lemastero) [reply src/test/scala/io/iohk/ethereum/proof/MptProofVerifier.scala]: Now it is ADT.

Comment (lemastero) [reply src/test/scala/io/iohk/ethereum/proof/MptProofVerifier.scala]: Thank for this remark 👍

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (AnastasiiaL) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala]: I was thinking the same, forgot to leave a comment :)

Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala]: Done

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala
Comment (lemastero) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala]: Done

**Comment by dzajkowski** [comment]
@AnastasiiaL @bsuieric would it be possible to bring this pr to a mergeable state and get it into dev? it's already large, further work can happen in separate PRs. WDYT?

**Comment by AnastasiiaL** [comment]
> @AnastasiiaL @bsuieric would it be possible to bring this pr to a mergeable state and get it into dev? it's already large, further work can happen in separate PRs. WDYT?

the PR is ready to be merged, waiting for approval

**Comment by bsuieric** [comment]
How can I approve the PR?

**Review by bsuieric — APPROVED** [review]

**Comment by bsuieric** [comment]
done

**Review by dzajkowski — APPROVED** [review]

**Pull request merged**
